### PR TITLE
one shot system cleanup

### LINKED ISSUE #16497: Clean up one-shot systems API
There are a few inconsistencies and gaps in the current one-shot systems API:

- There should be a `Commands::unregister_system_cached` method
- `Commands::` and `World::run_system_with_input` should be renamed to `run_system_with` for consistency
- `Commands::` and `World::run_system_once_with` should accept `(system, input)` in that order rather than `(input, system)` for consistency

### PR DESCRIPTION
**Pull request by pemattern**
# Objective

- Fixes #16497
- This is my first PR, so I'm still learning to contribute to the project

## Solution

- Added struct `UnregisterSystemCached` and function `unregister_system_cached`
- renamed `World::run_system_with_input` to `run_system_with`
- reordered input parameters for `World::run_system_once_with`

## Testing

- Added a crude test which registers a system via `World::register_system_cached`, and removes it via `Command::unregister_system_cached`.

## Migration Guide

- Change all occurrences of `World::run_system_with_input` to `World::run_system_with`.
- swap the order of input parameters for `World::run_system_once_with` such that the system comes before the input.

### CODE REVIEW

**Comment by pemattern** [comment]
added migration guide

File: crates/bevy_ecs/src/system/system_registry.rs
Comment (benfrankel) [inline crates/bevy_ecs/src/system/system_registry.rs:580]:
```suggestion
    _phantom: PhantomData<fn() -> (I, O, M)>,
```
Optional

File: crates/bevy_ecs/src/system/system_registry.rs
Comment (benfrankel) [inline crates/bevy_ecs/src/system/system_registry.rs:579]: Technically `world.unregister_system_cached` doesn't actually use the system argument passed into it, so this field could be removed. We could add a private `World::unregister_system_cached_inner::<S>` method that `World::unregister_system_cached` and this command both call into, using `S` directly instead of passing in a value of type `S`.

This isn't a big deal though so I'm not blocking on it.

File: crates/bevy_ecs/src/system/commands/mod.rs
Comment (benfrankel) [inline crates/bevy_ecs/src/system/commands/mod.rs:2500]: Could actually use this system ID to check that the entity it points to (`SystemId::entity()`) exists, and later that the entity no longer exists after unregister.

File: crates/bevy_ecs/src/system/system_registry.rs
Comment (chescock) [reply crates/bevy_ecs/src/system/system_registry.rs:580]: Another option is to remove the `IntoSystem` constraint from the type entirely and just have 

```rust
pub struct UnregisterSystemCached<S> {
    system: S,
}
```

You need the constraints on `fn unregister_system_cached` to have the `Command` impl, but I don't think you actually need them on the `struct`.

**Review by chescock — APPROVED** [review]

File: crates/bevy_ecs/src/system/commands/mod.rs
Comment (pemattern) [reply crates/bevy_ecs/src/system/commands/mod.rs]: Much better, thanks!

File: crates/bevy_ecs/src/system/commands/mod.rs
Comment (benfrankel) [inline crates/bevy_ecs/src/system/commands/mod.rs:2506]: The resource asserts would still be useful here alongside the entity asserts.

**Review by benfrankel — CHANGES_REQUESTED** [review]
Will approve after this change.

**Review by benfrankel — APPROVED** [review]

**Review by BenjaminBrienen — APPROVED** [review]

**Pull request merged**
